core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/batchutilities/BatchManager.java

```diff
@@ -0,0 +1,250 @@
+
+    /**
+     * A scheduled executor that periodically schedules {@link #flushBuffer}.
+     */
+    private final ScheduledExecutorService scheduledExecutor;
+
+    private BatchManager(Builder<RequestT, ResponseT, BatchResponseT> builder) {
+        BatchOverrideConfiguration overrideConfiguration = Validate.notNull(builder.overrideConfiguration, "Null override"
+                                                                                                             + "Configuration");
+        this.requestsAndResponsesMaps = new BatchingMap<>();
+        this.maxBatchItems = overrideConfiguration.maxBatchItems();
+        this.maxBatchOpenInMs = overrideConfiguration.maxBatchOpenInMs();
+        this.batchingFunction = Validate.notNull(builder.batchingFunction, "Null batchingFunction");
+        this.mapResponsesFunction = Validate.notNull(builder.mapResponsesFunction, "Null mapResponsesFunction");
+        this.batchKeyMapperFunction = Validate.notNull(builder.batchKeyMapperFunction, "Null batchKeyMapperFunction");
+        this.scheduledExecutor = overrideConfiguration.scheduledExecutor();
+    }
+
+    public static <RequestT, ResponseT, BatchResponseT> Builder<RequestT, ResponseT, BatchResponseT> builder() {
+        return new Builder<>();
+    }
+
+    /**
+     * Buffers outgoing requests on the client and sends them as batch requests to the service. Requests are batched together
+     * according to a batchKey and are sent periodically to the service as determined by {@link #maxBatchOpenInMs}. If the
+     * number of requests for a batchKey reaches or exceeds {@link #maxBatchItems}, then the requests are immediately flushed
+     * and the timeout on the periodic flush is reset.
+     * By default, messages are batched according to a service's maximum size for a batch request. These settings can be
+     * customized via the configuration.
+     *
+     * @param request the outgoing request.
+     * @return a CompletableFuture of the corresponding response.
+     */
+    public CompletableFuture<ResponseT> sendRequest(RequestT request) {
+        CompletableFuture<ResponseT> response = new CompletableFuture<>();
+        try {
+            String batchKey = batchKeyMapperFunction.getBatchKey(request);
+            requestsAndResponsesMaps.batchBufferByKey(batchKey, () -> scheduleBufferFlush(batchKey, maxBatchOpenInMs.toMillis(),
+                                                                                          scheduledExecutor))
+                                    .put(request, response);
+
+            int requestsNum = requestsAndResponsesMaps.get(batchKey).requestSize();
+            if (requestsNum > maxBatchItems) {
+                cancelScheduledFlushIfNeeded(batchKey);
+            }
+            return response;
+        } catch (Exception e) {
+            response.completeExceptionally(e);
+        }
+        return response;
+    }
+
+    private void cancelScheduledFlushIfNeeded(String batchKey) {
```

Comment:
since there is another flushBuffer method that does the flushing, would it make sense to maybe call it `manualFlushBuffer` (since it is overriding the scheduled buffer flush to flush when the number of requests > maxBatchItesms)?

Reply:
Yeah, sounds good

Reply:
Renamed method.